### DISCUSSION TITLE
Is it possible to create a multi-level directory structure?

### DISCUSSION THREAD

**Discussion by jkleve** [copier-org/copier#683] [Q&A]
In java, the directory structure matches the package. If there's a package `com.google.gson` then the directory structure for this package may be `src/main/java/com/google/gson`.  Is it possible to create this directory structure with something like `src/main/java/{{package}}`?

**Comment by pawamoy** [comment]
Are there always three items in your package path? If yes, you can create directories like this:

```
src/
    main/
        java/
            {{package.split(".")[0]}}/
                {{package.split(".")[1]}}/
                    {{package.split(".")[2]}}/
```

**Reply by jkleve** [reply]
Unfortunately not.

**Reply by pawamoy** [reply]
Then I see no other way than using a task. You generate the contents of the last directory into a temporary one, and add a task that creates the directory hierarchy and moves the generated contents from the temp dir to the final dir of your package sources.

```
src/
    main/
        java/
tmpsrc/
    ...  # files and folders
```

Rough task example:

```bash
# task.sh.jinja
src=src/main/java/{{package.replace(".", "/")}}
mkdir -p $src
mv tmpsrc/* $src
rm -rf tmpsrc
```

**Reply by jkleve** [reply]
Thanks @pawamoy for directing me here. The working task I went with (`create_package_directories.sh.jinja`)
```sh
#!/bin/env bash
set -e

# This copier task creates the java package directory structure
# and moves all source/test files into their respective directories.

templates_dir=templates
src_dir=src/main/{{ package.replace(".", "/") }}
test_dir=src/test/{{ package.replace(".", "/") }}

mkdir -p "$src_dir"
mkdir -p "$test_dir"

echo "Moving $templates_dir/main to $src_dir"
mv "$templates_dir"/main/* "$src_dir"/

echo "Moving $templates_dir/main to $test_dir"
mv "$templates_dir"/test/* "$test_dir"/

rm -rf "$templates_dir"
```

**Comment by yajo** [comment]
I'm not sure if this will work, but try this directory name:

    {{package.replace('.', '\u2044')}}

**Reply by pawamoy** [reply]
Trying to trick the file system? :smile: It would be amazing if it did haha

**Reply by yajo** [reply]
I just tested it and it doesn't work. 😁 

Well... it actually creates a file with those `/` inside the file name, which I guess will be a problem everywhere. But strangely it works.

However this could be one valid feature request. We could include a constant like `OS_SEP` (coming from `os.sep`) in all templates. Then we could support explicitly the case of one of those separators being found in a dir/file name.

**Reply by pawamoy** [reply]
I like the idea.

```
{{package.replace('.', copier_sep)}}
```

...with `copier_sep` being `'\u2044'` (or any other rarely used character), and Copier would know to treat them as actual OS separators and run mkdir with the -p option (the Python equivalent of course).

Wouldn't it be a bit complex to support this in file names as well though?

```
{{package.replace('.', copier_sep)}}.jinja
```

...would be rendered as, for example, `com/google/gson.jinja`, and Copier would then have to understand it has to split the basename from the rest in order to first create the directories, then the file inside. Well, there's probably an elegant way to handle all the cases.

**Reply by yajo** [reply]
Yes. @jkleve feel free to create a feature request if you feel like it's necessary.

**Reply by jkleve** [reply]
@yajo It's probably pretty low priority but I created [an issue](https://github.com/copier-org/copier/issues/692). I only plan to support linux for my uses.

**Comment by mszalbach** [comment]
I have used copier for Java project templating and solved it like this, maybe this helps.

I created a helper question in the copier.yml


```
package_name:
  type: str
  help: What is your package name?
  default: "{{group_id}}.{{artifact_id | replace('-','.')}}"

package_folder:
  type: str
  default: "{{package_name | replace('.','/')}}"
```


Create the folder structure src/main/java and and add a folder called {[package_folder}}.
Works for Linux and Windows.

**Reply by yajo** [reply]
Weird... last time I tried that didn't work. https://github.com/copier-org/copier/discussions/683#discussioncomment-2840465
